File: typescript/examples/evm_basics.ts
#!/usr/bin/env npx ts-node
// @ts-nocheck
/**
 * HyperEVM Example
 *
 * Shows how to use standard Ethereum JSON-RPC calls on Hyperliquid's EVM chain.
 *
 * Usage:
 *     export QUICKNODE_ENDPOINT="https://your-endpoint.hype-mainnet.quiknode.pro/YOUR_TOKEN"
 *     npx ts-node evm_basics.ts
 */

import { HyperliquidSDK } from '@quicknode/hyperliquid-sdk';

const ENDPOINT = process.env.QUICKNODE_ENDPOINT;

if (!ENDPOINT) {
  console.error("Set QUICKNODE_ENDPOINT environment variable");
  process.exit(1);
}

async function main() {
  // Single SDK instance — access everything through sdk.info, sdk.core, sdk.evm, etc.
  const sdk = new HyperliquidSDK(ENDPOINT);
  const evm = sdk.evm;

  console.log("=".repeat(50));
  console.log("HyperEVM (Ethereum JSON-RPC)");
  console.log("=".repeat(50));

  // Chain info
  console.log("\n1. Chain Info:");
  const chainId = await evm.chainId();
  const blockNum = await evm.blockNumber();
  const gasPrice = await evm.gasPrice();
  console.log(`   Chain ID: ${chainId}`);
  console.log(`   Block: ${blockNum}`);
  console.log(`   Gas Price: ${(Number(gasPrice) / 1e9).toFixed(2)} gwei`);

  // Latest block
  console.log("\n2. Latest Block:");
  const block = await evm.getBlockByNumber("latest");
  if (block) {
    console.log(`   Hash: ${block.hash?.slice(0, 20)}...`);
    console.log(`   Txs: ${(block.transactions || []).length}`);
  }

  // Check balance
  console.log("\n3. Balance Check:");
  const addr = "0x0000000000000000000000000000000000000000";
  const balance = await evm.getBalance(addr);
  console.log(`   ${addr.slice(0, 12)}...: ${(Number(balance) / 1e18).toFixed(6)} ETH`);

  console.log("\n" + "=".repeat(50));
  console.log("Done!");
  console.log("\nFor debug/trace APIs, use: new EVM(endpoint, { debug: true })");
}

main().catch(console.error);

File: python/examples/evm_basics.py
#!/usr/bin/env python3
"""
HyperEVM Example

Shows how to use standard Ethereum JSON-RPC calls on Hyperliquid's EVM chain.

Requirements:
    pip install hyperliquid-sdk

Usage:
    export ENDPOINT="https://your-endpoint.hype-mainnet.quiknode.pro/TOKEN"
    python evm_basics.py
"""

import os
import sys

from hyperliquid_sdk import HyperliquidSDK

ENDPOINT = os.environ.get("ENDPOINT") or os.environ.get("QUICKNODE_ENDPOINT")

if not ENDPOINT:
    print("Error: Set ENDPOINT environment variable")
    print("  export ENDPOINT='https://your-endpoint.hype-mainnet.quiknode.pro/TOKEN'")
    sys.exit(1)


def main():
    print("=" * 50)
    print("HyperEVM (Ethereum JSON-RPC)")
    print("=" * 50)

    # Single SDK instance - access everything through sdk.info(), sdk.core(), sdk.evm()
    sdk = HyperliquidSDK(ENDPOINT)
    evm = sdk.evm()

    # Chain info
    print("\n1. Chain Info:")
    chain_id = evm.chain_id()
    block_num = evm.block_number()
    gas_price = evm.gas_price()
    print(f"   Chain ID: {chain_id}")
    print(f"   Block: {block_num}")
    print(f"   Gas Price: {gas_price / 1e9:.2f} gwei")

    # Latest block
    print("\n2. Latest Block:")
    block = evm.get_block_by_number(block_num)
    if block:
        block_hash = block.get("hash", "?")
        txs = block.get("transactions", [])
        print(f"   Hash: {block_hash[:20]}...")
        print(f"   Txs: {len(txs)}")

    # Check balance
    print("\n3. Balance Check:")
    addr = "0x0000000000000000000000000000000000000000"
    balance = evm.get_balance(addr)
    print(f"   {addr[:12]}...: {balance / 1e18:.6f} ETH")

    print()
    print("=" * 50)
    print("Done!")
    print("\nFor debug/trace APIs, use: sdk.evm(debug=True)")


if __name__ == "__main__":
    main()
